File: ts/256/189.ts
// 数组翻转
function rotate(nums: number[], k: number): void {
  const n = nums.length
  k %= n
  if (k === 0) return
  const reverse = (l: number, r: number): void => {
    while (l < r) {
      let tmp = nums[l]
      nums[l] = nums[r]
      nums[r] = tmp
      l += 1
      r -= 1
    }
  }
  reverse(0, n - 1)
  reverse(0, k - 1)
  reverse(k, n - 1)
};

File: c/256/189.c
void swap(int* l, int* r) {
  int t = *l;
  *l = *r, *r = t;
}

void reverse(int* nums, int l, int r) {
  while (l < r) {
    swap(&nums[l], &nums[r]);
    l++;
    r--;
  }
}

void rotate(int* nums, int numsSize, int k) {
  k %= numsSize;
  if (k == 0) return;
  reverse(nums, 0, numsSize - 1);
  reverse(nums, 0, k - 1);
  reverse(nums, k, numsSize - 1);
}